File: suites/osprey-compute-kernels/reference/cases/collatz/collatz.hs
-- Collatz (3n+1) stopping time summed over 1..N — integer division (n/2) in deep recursion.
module Main (main) where

collatz :: Int -> Int
collatz 1 = 0
collatz n
  | even n    = 1 + collatz (n `div` 2)
  | otherwise = 1 + collatz (3 * n + 1)

step :: Int -> Int -> Int
step acc i = acc + collatz i

main :: IO ()
main = print (foldl step 0 [1 .. 100000])

File: suites/osprey-compute-kernels/reference/cases/collatz/collatz.rs
// Collatz (3n+1) stopping time summed over 1..N — integer division (n/2) in deep recursion.
fn collatz(n: i64) -> i64 {
    if n == 1 {
        0
    } else if n % 2 == 0 {
        1 + collatz(n / 2)
    } else {
        1 + collatz(3 * n + 1)
    }
}

fn main() {
    let mut acc: i64 = 0;
    let mut i: i64 = 1;
    while i < 100001 {
        acc += collatz(i);
        i += 1;
    }
    println!("{}", acc);
}
